src/software/ai/hl/stp/tactic/patrol_tactic_test.cpp

```diff
@@ -0,0 +1,289 @@
+
+    // Act and Assert
+
+    for (int j = 0; j < 2; j++)
+    {
+        for (int i = 0; i < patrolPoints.size(); i++)
+        {
+            std::shared_ptr<Action> action_ptr = tactic.getNextAction();
+            ASSERT_NE(nullptr, action_ptr);
+
+            std::shared_ptr<MoveAction> move_action =
+                std::dynamic_pointer_cast<MoveAction>(action_ptr);
+            compareMoveActions(expectedActions[i], move_action);
+
+            simulateActionToCompletion(robot, robotStateToCompleteActions[i], action_ptr,
+                                       tactic);
+        }
+    }
+}
+
+TEST_F(PatrolTacticTest, cost_of_non_assigned_robot_when_tactic_is_assigned)
+{
+    // setup
+    Point robotStartingPoint = Point(1, 1);
+    Vector initialVelocity   = Vector(0, 1);
+
+    Robot assigned_robot = Robot(0, robotStartingPoint, initialVelocity, Angle::zero(),
+                                 AngularVelocity::zero(), Timestamp::fromSeconds(0));
+
+    Robot non_assigned_robot =
+        Robot(1, robotStartingPoint, initialVelocity, Angle::zero(),
+              AngularVelocity::zero(), Timestamp::fromSeconds(0));
+
+    std::vector<Point> patrolPoints{Point(2, 3), Point(4, 3)};
+    World world = ::Test::TestUtil::createBlankTestingWorld();
+
+    PatrolTactic tactic = PatrolTactic(patrolPoints, 1.0, Angle::zero(), 2.0);
+
+    tactic.updateRobot(assigned_robot);
+
+    double cost = tactic.calculateRobotCost(non_assigned_robot, world);
+
+    EXPECT_EQ(COST_OF_NONASSIGNED_ROBOT_UNASSIGNED_TACTIC, cost);
+}
+
+TEST_F(PatrolTacticTest, cost_of_non_assigned_robot_when_tactic_is_not_assigned)
+{
+    // setup
+    Point robotStartingPoint = Point(1, 1);
+    Vector initialVelocity   = Vector(0, 1);
+
+    Robot non_assigned_robot =
+        Robot(1, robotStartingPoint, initialVelocity, Angle::zero(),
+              AngularVelocity::zero(), Timestamp::fromSeconds(0));
+
+    std::vector<Point> patrolPoints{Point(2, 2), Point(4, 3)};
+    World world = ::Test::TestUtil::createBlankTestingWorld();
+
+    PatrolTactic tactic = PatrolTactic(patrolPoints, 1.0, Angle::zero(), 2.0);
+
+    double cost = tactic.calculateRobotCost(non_assigned_robot, world);
+
+    EXPECT_NEAR(0.15, cost, 0.05);
+}
+
+TEST_F(PatrolTacticTest, cost_of_already_assigned_robot)
+{
+    // setup
+    int robotId              = 0;
+    Point robotStartingPoint = Point(1, 1);
+    Vector initialVelocity   = Vector(0, 1);
+
+    Robot assigned_robot =
+        Robot(robotId, robotStartingPoint, initialVelocity, Angle::zero(),
+              AngularVelocity::zero(), Timestamp::fromSeconds(0));
+
+    std::vector<Point> patrolPoints{Point(2, 3), Point(4, 3)};
+    World world = ::Test::TestUtil::createBlankTestingWorld();
+
+    PatrolTactic tactic = PatrolTactic(patrolPoints, 1.0, Angle::zero(), 2.0);
+
+    tactic.updateRobot(assigned_robot);
+
+    double cost = tactic.calculateRobotCost(assigned_robot, world);
+
+    EXPECT_EQ(COST_OF_ASSIGNED_ROBOT, cost);
+}
```

Comment:
EOF extra line needed here.

Reply:
In my ide there is a line. Isnt it only needed if the small red symbol appears?

Reply:
Should be fine then.